community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/InstanceSelector.java

```diff
@@ -46,33 +68,56 @@
      * @param slot slot number to return instance for.
      * @return the instance at the given {@code slot}.
      */
-    T select( int slot )
+    T select( IndexSlot slot )
     {
-        if ( instances[slot] == null )
+        if ( !instances.containsKey( slot )  )
         {
             throw new IllegalStateException( "Instance is not instantiated" );
         }
-        return instances[slot];
+        return instances.get( slot );
     }
 
     /**
-     * Convenience method typically for calling a method on each of the instances, a method that returns another type of instance.
-     * Even called on instances that haven't been instantiated yet. All those created instances are put into the provided array and returned.
+     * Map current instances to some other type using the converter function.
+     * Even called on instances that haven't been instantiated yet.
+     * Mapping is preserved in returned {@link EnumMap}.
      *
-     * @param target array to put the created instances into, also returned.
      * @param converter {@link ThrowingFunction} which converts from the source to target instance.
      * @param <R> type of returned instance.
      * @param <E> type of exception that converter may throw.
-     * @return the target array which was passed in, now populated.
+     * @return A new {@link EnumMap} containing the mapped values.
      * @throws E exception from converter.
      */
-    <R,E extends Exception> R[] instancesAs( R[] target, ThrowingFunction<T,R,E> converter ) throws E
+    <R,E extends Exception> EnumMap<IndexSlot,R> map( ThrowingFunction<T,R,E> converter ) throws E
     {
-        for ( int slot = 0; slot < instances.length; slot++ )
+        EnumMap<IndexSlot,R> result = new EnumMap<>( IndexSlot.class );
+        for ( IndexSlot slot : IndexSlot.values() )
         {
-            target[slot] = converter.apply( select( slot ) );
+            result.put( slot, converter.apply( select( slot ) ) );
         }
-        return target;
+        return result;
+    }
+
+    /**
+     * Map current instances to some other type using the converter function,
+     * without preserving the mapping.
+     * Even called on instances that haven't been instantiated yet.
+     *
+     * @param converter {@link ThrowingFunction} which converts from the source to target instance.
+     * @param <R> type of returned instance.
+     * @param <E> type of exception that converter may throw.
+     * @return A new {@link EnumMap} containing the mapped values.
+     * @throws E exception from converter.
+     */
+    @SuppressWarnings( "unchecked" )
+    <R,E extends Exception> Iterable<R> flatMap( ThrowingFunction<T,R,E> converter ) throws E
```

Comment:
its not really flat map since we not flatting anything here

Reply:
@MishaDemianenko Well, we "flat" away the mapping but I get your point. Do you have any suggestions?
- mapValues
- mapAsIterable
- simpleMap
- mapSimple
Hmm...

Reply:
what about "transform"?